### DISCUSSION TITLE
Docker deployment with nginx causes significant lag

### DISCUSSION THREAD

**Discussion by Mikailo** [neurodesk/neurodesk.github.io-hugo#355] [Q&A] [answered]
Hello,
In our current configuration you are running in a VM a docker version of neurodesk.
To access the vm normally you need a vpn. If we use the vpn and connect directly to the neurodesk instance there is no issue.

To make it accessible we also have an nginx server that connects to it. The issue is that when connecting through the nginx server, there is significant lag, both when inputting commands (i.e. clicking on anything takes 10-15 seconds) and when anything changes the page seems to freeze for a few seconds.

I checked both nginx server configuration if it was causing the issue but I haven't come up with anything yet.

Do you have any ideas what might be causing this and any solution?

Thank you

**Comment by stebo85** [comment]
This sounds like a suboptimal setup, but I don't quite understand how you configured it all. There might be a much easier setup that doesn't cause such issues. Would you like to organise a quick zoom session? https://calendly.com/s-bollmann/meeting

**Reply by Mikailo** [reply]
Yes that may be ideal, but I wouldn't want to take up your time unless necessary.
To be a bit more precise just in case it helps, the things that are truly relevant are the following:
 Neurodesk is running on docker in port: 8080 
If I access Neurodesk from <vm_ip>:8080 then everything runs smoothly

Now we also have an nginx server which redirects to neurodesk, with the following configuration. 
`
 server {
            listen 446 ssl;
            listen [::]:446 ssl;
            ssl_certificate /etc/nginx/certificate/nginx-certificate.crt;
            ssl_certificate_key /etc/nginx/certificate/nginx.key;
            server_name _;
            location / {
                proxy_pass <vm_ip>:8080;
            }
`

So now I can also access Neurodesk from <vm_ip>:446
The issue is that when I connect through the second address there is significant lag between each action happening, (i.e. click or writing anything)
Generally Neurodesk runs smoothly and fast the issue seems to be only between the connection from the user browser to Neurodesk. So it must be an issue with nginx configuration I assume. 

I hope our configuration and the issue is a bit clearer.

**Reply by stebo85** [reply]
So it looks like your nginx is not configured correctly. 

Things you could try:
- use nginx without ssl and see if the certificates are creating the problem
- you could use an SSH tunnel instead of nginx to connect to the VM - that would also be much securer and is what we generally recommend: https://www.neurodesk.org/docs/getting-started/neurodesktop/cloud/#1-connect-to-cloud-server
- we use nginx and certmanager within kubernetes to create our hosted installations of neurodesk, like https://bheurope.neurodesk.org/ - this is of course quite a complex setup ... but if you need something like this then kubernetes might be an option

**Reply by Mikailo** [reply]
So indeed I only removed sll and the lag is mostly gone. A tiny bit remains, but its mostly negligible.
I am not that familiar with nginx so I can't imagine what would cause this?

Is it possible to send me your nginx configuraiton, the one you mentioned with kubernetes in case this can help?

Thanks a lot for the assistance

**Reply by aswinnarayanan** [reply]
Hi @Mikailo, would you be able to provide some description of what you are trying to set up?

Are you running a Neurodesk docker container on one machine, and wanting to access the desktop from a different machine?
Is Neurodesk and nginx running on the same machine? 
Are you accessing from the same network? Or outside that network? 
And does the desktop need to be used by a single user? Or multiple users?

Please note:
- If you were only able to connect to the Neurodesk instance only via a VPN, and nginx is helping bypass the VPN requirement, there is a possibility that nginx could be providing an insecure solution, or broadcasting your server onto the web via an open port on the firewall
- 446 is a non-standard port for https/SSL. And 1-1023 port numbers are usually not-recommended for non-standard work

**Reply by Mikailo** [reply]
Hello,
I have created a simplified diagram with only the relevant components to better explain, what we want and what we have tried.
![Neurodesk Explanation](https://github.com/NeuroDesk/neurodesk.github.io/assets/37148156/f21055d3-1645-4fac-a907-48e8161deb94)

We are trying to host neurodesk to a VM we have access. To connect to the VM normally you need a vpn. But the nginx is also connected to a web hosting service, so it accessible normally through any browser without any vpn. This is the "Client Connection" in the diagram which we want eventually to work. So anyone connected to the internet with credentials can connect normally.

The rest of the connections are for setting up Neurodesk up and testing.

"Docker Port Connection 2" -> Isn't possible because I cant access the docker port without the VPN. 

"Docker Port Connection 1" -> When connected to the VPN "Client 1" can access Neurodesk directly through the Docker port. This is working fine without any issue.

"Nginx Port Connection" -> Our nginx is working as a reverse proxy and forwards requests to Neurodesk, as mentioned in my previous comment. When connected to the VPN "Client 1" can access Neurodesk directly through the Nginx port. Here the lagging issue I mentioned appears first. Per @stebo85 I disabled the ssl in the nginx and this essentially solved the issue.

"Client Connection" -> Here the application is also lagging but I can't disable the ssl because without it the web hosting doesn't allow us to connect to it.

So as far as I can tell the issue appears first when accessing it through nginx. That's where the problem first appears.

(You are right about the port, we did it because we are hosting multiple applications in the same VM, we should definitely change that, but I assume it's not the cause in this case)

I hope this also answers all your questions. Thank you again!

**Reply by stebo85** [reply, marked as answer]
Dear @Mikailo,

A quick zoom call would be helpful to discuss your intentions and possible options to do this in a more secure and reliable way. Feel free to grab a timeslot from here: https://calendly.com/s-bollmann/meeting

Thank you
Steffen

**Reply by Mikailo** [reply]
Hello everyone,
So I finally figured out the issue and the cause and solution was much simpler.
The only part of our configuration that mattered in the end was the nginx server that was used as a reverse proxy and the neurodesktop docker.

The issue was in the nginx server settings were according to apache guacamole configuration (https://guacamole.apache.org/doc/gug/reverse-proxy.html) : a missing "proxy_buffering off" setting in the nginx configuration. This essentially solved all the connection issues and lagging.

I  think setting up a reverse proxy server is not that rare of a case, so it might be useful to add this information somewhere, if it doesn't exist. 

Again, thanks for your help!

**Reply by stebo85** [reply]
Dear @Mikailo - great to hear you got it to work! I added this to our documentation now.
